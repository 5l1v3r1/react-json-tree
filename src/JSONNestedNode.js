// @flow
import React from 'react';
import JSONArrow from './JSONArrow';
import getCollectionEntries from './utils/getCollectionEntries';
import JSONNode from './JSONNode';
import ItemRange from './ItemRange';
import createItemPreviewText from './utils/createItemPreviewText';
import createItemTypeText from './utils/createItemTypeText';
import isExpandable from './utils/isExpandable';

/**
 * Renders nested values (eg. objects, arrays, lists, etc.)
 */

import type {
  NestedType,
  RenderItemPreview,
  RenderLabel,
  Sorter,
  KeyPath,
  ShouldExpandNode,
  IsCustomNode,
  RenderValue,
  PostprocessValue
} from './types';
import type { StylingFunction } from 'react-base16-styling';

type DefaultProps = {
  circularCache: mixed[],
  level: number,
  collectionLimit: number
};

type Props = DefaultProps & {
  value: any,
  renderItemPreview: RenderItemPreview,
  nodeType: NestedType,
  hideRoot: boolean,
  styling: StylingFunction,
  collectionLimit: number,
  keyPath: KeyPath,
  renderLabel: RenderLabel,
  shouldExpandNode: ?ShouldExpandNode,
  sortObjectKeys: Sorter<string> | boolean,
  isCircular?: boolean,
  isCustomNode: IsCustomNode,
  renderValue: RenderValue,
  shouldExpandNode: ShouldExpandNode,
  postprocessValue: PostprocessValue
};

type State = {
  expanded: boolean
};

function renderChildNodes(props, from, to) {
  const {
    nodeType,
    value,
    collectionLimit,
    circularCache,
    keyPath,
    postprocessValue,
    sortObjectKeys,
    styling,
    isCustomNode,
    renderItemPreview,
    renderLabel,
    renderValue,
    shouldExpandNode
  } = props;
  const childNodes = [];

  const collectionEntries = getCollectionEntries(
    nodeType,
    value,
    sortObjectKeys,
    collectionLimit,
    from,
    to
  );

  collectionEntries.forEach(entry => {
    if (entry.isRange) {
      childNodes.push(
        <ItemRange
          styling={styling}
          nodeType={nodeType}
          key={`ItemRange--${entry.from}-${entry.to}`}
          from={entry.from}
          to={entry.to}
          renderChildNodes={(from, to) => renderChildNodes(props, from, to)}
        />
      );
    } else {
      const { key, value } = entry;
      const isCircular = circularCache.indexOf(value) !== -1;

      const node = (
        <JSONNode
          styling={styling}
          {...{
            postprocessValue,
            collectionLimit,
            isCustomNode,
            renderItemPreview,
            renderLabel,
            isCircular,
            renderValue,
            shouldExpandNode,
            sortObjectKeys
          }}
          key={`Node--${key}`}
          keyPath={[key, ...keyPath]}
          value={postprocessValue(value)}
          circularCache={[...circularCache, value]}
          hideRoot={false}
        />
      );

      if (node !== false) {
        childNodes.push(node);
      }
    }
  });

  return childNodes;
}

function getStateFromProps(props) {
  // calculate individual node expansion if necessary
  const expanded = props.shouldExpandNode && !props.isCircular
    ? props.shouldExpandNode(props.keyPath, props.value, props.level)
    : false;
  return {
    expanded
  };
}

export default class JSONNestedNode
  extends React.Component<DefaultProps, Props, State> {
  state: State;

  static defaultProps = {
    circularCache: [],
    level: 0,
    collectionLimit: 0
  };

  constructor(props: Props) {
    super(props);
    this.state = getStateFromProps(props);
  }

  componentWillReceiveProps(nextProps: Props) {
    const nextState = getStateFromProps(nextProps);
    if (getStateFromProps(this.props).expanded !== nextState.expanded) {
      this.setState(nextState);
    }
  }

  shouldComponentUpdate(nextProps: Props, nextState: State) {
    return !!Object.keys(nextProps).find(
      key =>
        key !== 'circularCache' &&
        (key === 'keyPath'
          ? nextProps[key].join('/') !== this.props[key].join('/')
          : nextProps[key] !== this.props[key])
    ) || nextState.expanded !== this.state.expanded;
  }

  render() {
    const {
      renderItemPreview,
      nodeType,
      value,
      hideRoot,
      styling,
      collectionLimit,
      keyPath,
      renderLabel,
      level
    } = this.props;
    const { expanded } = this.state;
    const renderedChildren = expanded || (hideRoot && level === 0)
      ? renderChildNodes({ ...this.props, level: level + 1 })
      : null;

    const itemTypeText = createItemTypeText(nodeType);
    const itemPreviewText = createItemPreviewText(
      nodeType,
      value,
      collectionLimit
    );
    const renderedItemPreview = renderItemPreview(
      nodeType,
      value,
      itemTypeText,
      itemPreviewText
    );
    const expandable = isExpandable(nodeType, value);
    const stylingArgs = [keyPath, nodeType, expanded, expandable];

    return hideRoot
      ? <li {...styling(['rootNode', 'rootNodeColor'], ...stylingArgs)}>
          <ul
            {...styling(
              ['rootNodeChildren', 'rootNodeChildrenColor'],
              ...stylingArgs
            )}
          >
            {renderedChildren}
          </ul>
        </li>
      : <li {...styling(['nestedNode', 'nestedNodeColor'], ...stylingArgs)}>
          {expandable &&
            <JSONArrow
              arrowStyle="single"
              styling={styling}
              nodeType={nodeType}
              expanded={expanded}
              onClick={this.handleClick}
            />}
          <label
            {...styling(
              [
                'label',
                'labelColor',
                'nestedNodeLabel',
                'nestedNodeLabelColor'
              ],
              ...stylingArgs
            )}
            onClick={expandable && this.handleClick}
          >
            {renderLabel(...stylingArgs)}
          </label>
          <span
            {...styling(
              ['nestedNodeItemPreview', 'nestedNodeItemPreviewColor'],
              ...stylingArgs
            )}
            onClick={expandable && this.handleClick}
          >
            {renderedItemPreview}
          </span>
          <ul
            {...styling(
              ['nestedNodeChildren', 'nestedNodeChildrenColor'],
              ...stylingArgs
            )}
          >
            {renderedChildren}
          </ul>
        </li>;
  }

  handleClick = () => this.setState({ expanded: !this.state.expanded });
}
