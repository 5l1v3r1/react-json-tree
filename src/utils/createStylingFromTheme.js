// @flow
import { createStyling } from 'react-base16-styling';
import jsontree from '../themes/jsontree';

const colorMap = theme => ({
  BACKGROUND_COLOR: theme.base00,
  TEXT_COLOR: theme.base07,
  STRING_COLOR: theme.base0B,
  DATE_COLOR: theme.base0B,
  NUMBER_COLOR: theme.base09,
  BOOLEAN_COLOR: theme.base09,
  NULL_COLOR: theme.base08,
  UNDEFINED_COLOR: theme.base08,
  FUNCTION_COLOR: theme.base08,
  SYMBOL_COLOR: theme.base08,
  LABEL_COLOR: theme.base0D,
  ARROW_COLOR: theme.base0D,
  ITEM_STRING_COLOR: theme.base0B,
  ITEM_STRING_EXPANDED_COLOR: theme.base03
});

const valueColorMap = colors => ({
  String: colors.STRING_COLOR,
  Date: colors.DATE_COLOR,
  Number: colors.NUMBER_COLOR,
  Boolean: colors.BOOLEAN_COLOR,
  Null: colors.NULL_COLOR,
  Undefined: colors.UNDEFINED_COLOR,
  Function: colors.FUNCTION_COLOR,
  Symbol: colors.SYMBOL_COLOR
});

const getDefaultThemeStyling = theme => {
  const colors = colorMap(theme);

  return {
    tree: {
      border: 0,
      padding: 0,
      margin: 0,
      listStyle: 'none',
      MozUserSelect: 'none',
      WebkitUserSelect: 'none',
      fontFamily: 'inconsolata, monospace'
    },

    treeColor: {
      backgroundColor: colors.BACKGROUND_COLOR
    },

    value: ({ style }, nodeType, keyPath) => ({
      style: {
        ...style,
        paddingTop: '0.25em',
        paddingLeft: keyPath.length > 1 ? '2.125em' : '1.25em',
        paddingRight: 0,
        marginLeft: '0.875em',
        WebkitUserSelect: 'text',
        MozUserSelect: 'text',
        wordWrap: 'break-word',
        textIndent: '-0.5em',
        wordBreak: 'break-all',
        whiteSpace: 'pre-wrap'
      }
    }),

    label: {
      display: 'inline-block'
    },

    labelColor: {
      color: colors.LABEL_COLOR
    },

    valueLabel: {
      margin: '0 0.5em 0 0'
    },

    valueTextColor: ({ style }, nodeType) => ({
      style: {
        ...style,
        color: valueColorMap(colors)[nodeType]
      }
    }),

    itemRange: ({ style }, expanded) => ({
      style: {
        ...style,
        paddingTop: expanded ? 0 : '0.25em',
        cursor: 'pointer'
      }
    }),

    itemRangeColor: {
      color: colors.LABEL_COLOR
    },

    arrow: ({ style }, nodeType, expanded) => ({
      style: {
        ...style,
        marginLeft: 0,
        transition: '150ms',
        WebkitTransition: '150ms',
        MozTransition: '150ms',
        WebkitTransform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        MozTransform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        transform: expanded ? 'rotateZ(90deg)' : 'rotateZ(0deg)',
        transformOrigin: '45% 50%',
        WebkitTransformOrigin: '45% 50%',
        MozTransformOrigin: '45% 50%',
        position: 'relative',
        lineHeight: '1.1em',
        fontSize: '0.75em'
      }
    }),

    arrowInner: {
      position: 'absolute',
      top: 0,
      left: '-0.4em'
    },

    arrowContainer: ({ style }, arrowStyle) => ({
      style: {
        ...style,
        display: 'inline-block',
        paddingRight: '0.5em',
        paddingLeft: arrowStyle === 'double' ? '1em' : 0,
        cursor: 'pointer'
      }
    }),

    arrowSignColor: {
      color: colors.ARROW_COLOR
    },

    nestedNode: ({ style }, keyPath, nodeType, expanded, expandable) => ({
      style: {
        ...style,
        position: 'relative',
        paddingTop: keyPath.length > 1 ? '0.25em' : 0,
        marginLeft: keyPath.length > 1 ? '0.875em' : 0,
        paddingLeft: !expandable ? '1.125em' : 0
      }
    }),

    rootNode: {
      padding: 0,
      margin: 0
    },

    nestedNodeLabel: ({ style }, keyPath, nodeType, expanded, expandable) => ({
      style: {
        ...style,
        margin: 0,
        padding: 0,
        WebkitUserSelect: expandable ? 'inherit' : 'text',
        MozUserSelect: expandable ? 'inherit' : 'text',
        cursor: expandable ? 'pointer' : 'default'
      }
    }),

    nestedNodeItemPreview: {
      paddingLeft: '0.5em',
      cursor: 'default'
    },

    nestedNodeItemPreviewColor: ({ style }, keyPath, nodeType, expanded) => ({
      style: {
        ...style,
        color: expanded
          ? colors.ITEM_STRING_EXPANDED_COLOR
          : colors.ITEM_STRING_COLOR
      }
    }),

    nestedNodeChildren: ({ style }, nodeType, expanded) => ({
      style: {
        ...style,
        padding: 0,
        margin: 0,
        listStyle: 'none',
        display: expanded ? 'block' : 'none'
      }
    }),

    rootNodeChildren: {
      padding: 0,
      margin: 0,
      listStyle: 'none'
    }
  };
};

export default createStyling(getDefaultThemeStyling, {
  defaultBase16: jsontree,
  base16Themes: { jsontree }
});
