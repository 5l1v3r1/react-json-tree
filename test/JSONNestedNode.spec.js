// @flow
import React from 'react';
import test from 'ava';
import renderer from 'react-test-renderer';

import JSONNestedNode from '../src/JSONNestedNode';

import createStylingFromTheme from '../src/utils/createStylingFromTheme';

const styling = createStylingFromTheme(undefined);

test('should render JSONNestedNode', t => {
  const tree = (
    <JSONNestedNode
      value={[]}
      styling={styling}
      nodeType="Array"
      hideRoot
      collectionLimit={10}
      keyPath={['root']}
      renderLabel={() => {}}
      sortObjectKeys
      isCustomNode={() => false}
      renderValue={() => {}}
      shouldExpandNode={() => false}
      postprocessValue={v => v}
      renderItemPreview={() => {}}
    />
  );

  t.snapshot(renderer.create(tree).toJSON());
});
