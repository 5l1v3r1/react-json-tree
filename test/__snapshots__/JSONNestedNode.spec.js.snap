exports[`should render JSONNestedNode 1`] = `"{\"__ava_react_jsx\":{\"type\":\"li\",\"props\":{\"style\":{\"padding\":0,\"margin\":0}},\"children\":[{\"type\":\"ul\",\"props\":{\"style\":{\"padding\":0,\"margin\":0,\"listStyle\":\"none\"}},\"children\":null}]}}"`;
